articles/grand-rename.md

```diff
@@ -0,0 +1,97 @@
+
+But over time, we added more and more SDKs and now we have 5, but only 4 of them are also called "SDK": The initial Swift client was simply called "SwiftClient". We used the platform-specific terminology for things, leading to different names in some places. And we also learned the power of grouping signal names and event parameters as documented in our [naming guideline](https://telemetrydeck.com/docs/articles/signal-type-naming/). Because we sort everything alphabetically in our UI, it's a natural help in finding things more quickly if related things have the same prefix. Also, users could accidentally use the same signal or parameter name and override ours, causing confusion in the insight data.
+
+After more than 3 years working on TelemetryDeck, we think it's about time to streamline things for a more clear long-term future.
+
+## Consistency, Longevity, and Clarity
+
+Our main goals with our new naming scheme were to be internally consistent, but at the same time make sure the names are abstract enough to work flexibly for changes to come in the foreseeable future. Too abstract names can be confusing and cumbersome, so we additionally strived for clarity.
+
+The most obvious change was to rename our Swift repository from "SwiftClient" to "SwiftSDK". Thanks to GitHub redirections, this change should have no effect on our existing customers, but long term, it should make things more consistent. 🎉
+
+For our signals & parameters, we decided the prefix them with `TelemetryDeck` to make the source clear and avoid amgiguity. Additionally, we analyzed all the signals and parameters we currently send and future ones we have in our roadmap and grouped them as follows:
+
+### Signals
+
+- `Session`: Anything related to starting, pausing, continuing, or ending a user session.
+
+### Parameters
+
+- `AppContext`: The context the app runs in, such as the locale, the debug mode, or the target environment.
+- `AppInfo`: Information about the specific app build, such as version, build number, or SDKs compiled with.
+- `Device`: All about the device running the application, such as operating system, model name, or architecture.
+- `SDK`: Information about the TelemetryDeck SDK, such as the version number or supported features.
+
+{% noteinfo "No action needed" %}
+To make sure these changes don't affect existing customers with existing insights anytime soon, we decided to keep sending the old names for a transition period of at least 1 year while also sending the new ones. We have also auto-migrated any insights that were using the old names in their filters to work with the new system. This was safely achieved by replacing filters like "platform == iOS" with something like "platform == iOS OR TelemetryDeck.SystemInfo.platform == iOS" accepting both styles, ensuring you will continue to see historic data from before the transition, during the transition, and after the transition.
+{% endnoteinfo %}
```

Comment:
I'd suggest not promising auto-migration right now. Instead, let's promise that existing SDKs (we might call it SwiftSDK 2.0.0) will send new and deprecated parameters, but SwiftSDK 3.0.0 will drop the deprecated parameters when we release it in a year. 

In the coming year, we can then present a sort of "Migration Wizard" that will try to automatically update filters when it runs, and also list GroupBy Queries and similar insights that might need to be manually updated. 

We can either run this Wizard on the server in the background, if we're sure we can auto-migrate everything, or we can run it on the frontend with user interaction if we think some user consent is required.

Reply:
I can imagine some folks may ask if sending both signal types would also mean they'd be billed as two signals?

Not sure how the TD backend works, so I'm thinking why task the client with the migration? Perhaps we could find a way to make it so existing apps don't have to migrate right away but new apps can adopt the new signal types. And then v3 will no longer allow sending deprecated signal types.

For example, in v2.0, we extend the configuration object with a boolean parameter like `TelemetryDeck.useModernSignalTypes` set to `false` by default. This will be included in the additional payload which the server can use to rename the parameters during ingestion.

Users are welcome to set this to `true` in order to adopt the new signal types early during v2 (e.g. for new apps). When v3 is launched, the value is set to true and can no longer be overwritten.

Reply:
Alright, I will make the adjustment, seems like a good approach as well! 👍

Reply:
By the way, if you want to make a major new release of the SDK for this rename, I would change my opinion on the scope of this PR and include the consistency of the calls across al SDKs. My main problem with them is that in the docs, we don't have the option to write the correct calls for every SDK, so we have to pick one. But if the APIs all were named consistently, we could simply write one piece of code in the docs, and the other platforms would look very similar.

I just had a great discussion with @kkostov on Discord (we happened to meet today in the channel!) and we agreed that for Swift, Kotlin, and Flutter renaming `TelemetryManager.send(:with:)` to `TelemetryDeck.signal(:parameters:)` and additionally streamlining also the `requestImmediateSync()` approach introduced in https://github.com/TelemetryDeck/SwiftClient/pull/136 (rather than having two separate functions in Android) would be better.

We would also check if the initialization code has differently named APIs and adjust that, too. It should be totally possible to make these renaming changes without breaking existing code (via overloads and maybe deprecations).

What do you think? @winsmith @timidak

Reply:
@kkostov Oh, I'm just seeing your comment now, it seems we were writing at the same time and GitHub didn't indicate in any way that there was a new comment when I sent mine.

I did think about your approach, but firstly the duplicate parameters won't count as a new signal, so users shouldn't be concerned. The one extra signal we will send for the `TelemetryDeck.Session.started` on app start shouldn't make a big difference as most apps send multiple signals during a session and the percentage of one extra signal per session becomes relatively low.

If anything, I would suggest that we give users the option to opt-out of the deprecated calls and exclusively send the new ones. Users just need to make sure they update their insights to the new ones if they have any in those cases and then they wouldn't have any additional signals. But like I said above, I'm not even sure if this is necessary. If we were to migrate all the Android lifecycle events over, then you would be right, but I'm assuming they are going to get removed soon (maybe alongside the 2.0 migration?). I can imagine they produce a lot of signals, like each time a user looks up something to answer the questions in my crossword puzzle game and and return to the app, it's at least 2 signals for every lookup which can lead to >100 signals per user per puzzle.

Reply:
@winsmith I adjusted the note as you requested in 8756fd592da067913ea5cbb96b61ce3670448951. I didn't mention any version details, but tried to make clear we are aware of potential migration needs and will do our best to support existing customers in the migration in some way. I think the most important info here is the title: "No immediate action needed".

Reply:
Calling it `TelemetryDeck.signal` would be in sync with the JavaScript SDK and our React implementation. So I'd be super okay with that. If I get this correctly `TelemetryDeck.record` would be the same as `TelemetryDeck.queue` in the JavaScriptSDK (cf. [Code in the JsSDK](https://github.com/TelemetryDeck/JavaScriptSDK/blob/1e16779aefe037ddf24a5cef0290d9a6a2bbce12/src/telemetrydeck.js#L75)). 

I personally like the name queue, but I don't have a strong opinion which name might fit better. The other concept sounds to me like the `flush` method in the JavaScript SDK (refer to [this line](https://github.com/TelemetryDeck/JavaScriptSDK/blob/1e16779aefe037ddf24a5cef0290d9a6a2bbce12/src/telemetrydeck.js#L89)).

In general, I don't have a strong opinion, but I can see that naming consistency across SDKs should be addressed.

Reply:
@timidak What's the difference between `signal` and `queue`? And why aren't you using a verb like `enqueue`?

In Swift, the `signal` method is enqueueing sending the signal by default. There's no API to do it synchronously, so the `signal` is basically an `enqueue` by default. We avoid a synchronous signal to reduce server load. Same for what you call `flush` (is this a term of art? never heard it and on first thought I assumed it deletes all enqueued signals even if not sent yet), to reduce server load I've made sure that it only "directly" sends all enqueued signals if there was no call to "flush" within the last 10 seconds. Therefore the `request` prefix in `requestImmediateSync()`.

But I guess things might be a little different on the web. We should only make consistent, what has a similar behavior from the users perspective. If a synchronous call is the default on Frontend and what people should use, it can make sense to keep the name `signal` which is asynchronous on Mobile. If `flush` also continues to behave differently, it doesn't have to have the same name. But I would suggest `immediateSync` for consistency even if you don't implement the check for "already called in last 10 seconds", which might not be possible in Frontend, I don't know.

Reply:
Background info – Here's how we've been using these words so far, at least to my memory:

- send: Either sends a signal directly, or enqueues is, depending on SDK
- queue, enqueue: Add signal to a queue to be sent later (this is not  to reduce server load, but mostly to reduce antenna activation times for mobile devices)
- flush: send all signals that are in the queue right now.

Reply:
> By the way, if you want to make a major new release of the SDK for this rename, I would change my opinion on the scope of this PR and include the consistency of the calls across all SDKs. My main problem with them is that in the docs, we don't have the option to write the correct calls for every SDK, so we have to pick one. But if the APIs all were named consistently, we could simply write one piece of code in the docs, and the other platforms would look very similar if not exactly the same.

Streamlining is a great idea, let's do this.

Reply:
I'm all for this new nomenclature proposed by @Jeehut:

- Main singletons or namespaces are `TelemetryDeck` not `TelemetryManager`
- The verb for sending a signal is `signal`
- send addition parameters with the `parameters` paramete 
- It's not a payload, its `parameters`

Reply:
@winsmith If the only reason we were using a queue on mobile devices is to reduce battery drain, then I shouldn't have bothered to write the logic for `requestImmediateSync()` and just made it a simple `syncImmediately()`. I really thought it's about server load. 😅

Reply:
> The verb for sending a signal is signal

So just to be clear @winsmith @Jeehut, the Grand Rename will result in dropping `queue` method and repurposing `send` to provide the same functionality (appending signals to the cache), correct?

Note: In the Kotlin SDK the existing `queue` method drops signals older than 24h. The mindset at the time was that "important, don't drop it" signals would go out using `send` and others using `queue`.

Note2: The Flutter SDK exports a `send` method which routes to `send` on iOS and `queue` on Android.

Reply:
> So just to be clear @winsmith @Jeehut, the Grand Rename will result in dropping send method and repurposing queue as signal to provide the same functionality (appending signals to the cache), correct?

"Dropping" sounds like an immediate breaking change, but it's correct in general. In the Swift SDK I'm gonna keep all existing methods and just mark them as deprecated while they'll continue to work. But a fix-it is provided (or at least a message) that helps with the migration and no app using the SDK today will run into build errors, but just warnings. Until we decide it's time to fully drop them.

> Note: In the Kotlin SDK the existing queue method drops signals older than 24h. The mindset at the time was that "important, don't drop it" signals would go out using send and others using queue.

I'm not aware of any signal dropping in Swift, I also wouldn't expect that to happen and given that the queue will be the only option going forward, you should probably turn off the "drop after 24 h".

> Note2: The Flutter SDK exports a send method which routes to send on iOS and queue on Android.

That sounds reasonable as the old "send" is similar to "queue" on Android. Both will be "signal" in the future though, and I just created the PR for the future Swift SDK in https://github.com/TelemetryDeck/SwiftClient/pull/153 which contains the renaming changes in commit https://github.com/TelemetryDeck/SwiftClient/pull/153/commits/b6c7d436987bce3f5f8d64c74cf023d925844ba2.

Reply:
Please note that I found a `for` parameter and also a `floatValue` parameter on the `send` function. I renamed `for` to `customUserID` to be consistent with the configurations `defaultUserIdentifier` (which could be renamed to `defaultUserID` in the future). The `floatValue` didn't change. But I found that the order was like this:

1. Signal Name (required)
2. Custom User ID
3. Float Value
4. Parameters

It's more common to place parameters used more often at the front though and ones that are usually skipped to fall back to the default values at the end. So I changed the order as follows:

1. Signal Name (required)
2. Parameters
3. Float Value
4. Custom User ID

To still provide fix-its for most case, I added a copy of the function which only takes the Signal Name and Parameters and made sure it's favored over the function with more parameters if only 2 are provided. Otherwise, Xcode is unable to apply fix-its when the order of parameters change, but I think fix-its can be really useful to save time in migration.

Reply:
@kkostov I'm not gonna edit any of the Android / Flutter Setup documents, I think it's safer if you do the changes as soon as you've worked on the rename.